pkg/comp-functions/functions/vshnmariadb/user_management.go

```diff
@@ -0,0 +1,266 @@
+		svc.AddResult(runtime.NewWarningResult(fmt.Sprintf("cannot get userpassword from secret: %s", err)))
+	}
+
+	compositeCD := svc.GetConnectionDetails()
+
+	// url := getPostgresURLCustomUser(compositeCD, string(compositeCD["MARIADB_HOST"]), username)
+	url := fmt.Sprintf("mysql://%s:%s@%s:%s", username, userpassCD["userpass"], compositeCD["MARIADB_HOST"], compositeCD["MARIADB_PORT"])
+
+	om := metav1.ObjectMeta{
+		Name:      comp.GetLabels()["crossplane.io/claim-name"] + "-" + username,
+		Namespace: comp.GetClaimNamespace(),
+	}
+	if connectionDetailRef != nil {
+		om.Name = connectionDetailRef.Name
+		om.Namespace = connectionDetailRef.Namespace
+	}
+
+	userpassSecret := &corev1.Secret{
+		ObjectMeta: om,
+		Type:       corev1.SecretType("connection.crossplane.io/v1alpha1"),
+		Data: map[string][]byte{
+			"MARIADB_USERNAME": []byte(username),
+			"MARIADB_PASSWORD": userpassCD["userpass"],
+			"MARIADB_DB":       []byte(dbname),
+			"MARIADB_HOST":     compositeCD["MARIADB_HOST"],
+			"MARIADB_PORT":     compositeCD["MARIADB_PORT"],
+			"MARIADB_URL":      []byte(url),
+			"ca.crt":           compositeCD["ca.crt"],
+		},
+	}
+
+	err = svc.SetDesiredKubeObject(userpassSecret, fmt.Sprintf("%s-user-%s", comp.GetName(), username))
+	if err != nil {
+		svc.Log.Error(err, "cannot get userpassword from secret")
+		svc.AddResult(runtime.NewWarningResult(fmt.Sprintf("cannot get userpassword from secret: %s", err)))
+	}
+}
+
+func addProviderConfig(comp *vshnv1.VSHNMariaDB, svc *runtime.ServiceRuntime) {
+	cd := svc.GetConnectionDetails()
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "provider-conf-credentials",
+			Namespace: comp.GetInstanceNamespace(),
+		},
+		Data: map[string][]byte{
+			"username": cd["MARIADB_USERNAME"],
+			"password": cd["MARIADB_PASSWORD"],
+			"endpoint": cd["MARIADB_HOST"],
+			"port":     cd["MARIADB_PORT"],
+		},
+	}
+
+	err := svc.SetDesiredKubeObject(secret, comp.GetName()+"-provider-conf-credentials",
+		runtime.KubeOptionProtects("namespace-conditions"),
+		runtime.KubeOptionProtects("cluster"),
+		runtime.KubeOptionProtects(comp.GetName()+"-netpol"))
+	if err != nil {
+		svc.AddResult(runtime.NewWarningResult(fmt.Sprintf("cannot set credential secret for provider-sql: %s", err)))
+		svc.Log.Error(err, "cannot set credential secret for provider-sql")
+	}
+
+	tls := "preferred"
+	if comp.Spec.Parameters.TLS.TLSEnabled {
+		// We need to skip-verify here as with the postgresql.
+		// provider-sql does not support custom CA certs.
+		tls = "skip-verify"
+	}
+
+	config := &my1alpha1.ProviderConfig{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: comp.GetName(),
+		},
+		Spec: my1alpha1.ProviderConfigSpec{
+			TLS: &tls,
+			Credentials: my1alpha1.ProviderCredentials{
+				Source: "MySQLConnectionSecret",
+				ConnectionSecretRef: &xpv1.SecretReference{
+					Name:      "provider-conf-credentials",
+					Namespace: comp.GetInstanceNamespace(),
+				},
+			},
+		},
+	}
```

Comment:
I assume it requires specific field names and this is why we don't reuse existing secrets?

Reply:
Yes, the secret has to look exactly like this.